File: ui/src/lib/license/licenseCache.ts
/**
 * On-disk license cache with a 7-day TTL.
 *
 * Why disk? Per §3.2.1 of the master plan: a license-server outage MUST NOT
 * bring the customer's UI down. The Next.js server reads the cache on boot
 * (and on each `refresh()` cycle) so a transient outage degrades to
 * "stale-active" rather than "FOSS". After 7 days without a successful fetch
 * we give up and fall back to FOSS so a permanently-misconfigured server
 * doesn't pretend forever.
 *
 * Path: `~/.subscription-service/license-cache.json` by default. Tests pass an
 * explicit `cachePath` so they can write to a tmp dir.
 */

import { promises as fs } from "node:fs";
import path from "node:path";
import os from "node:os";
import type { EntitlementResponse, LicenseCacheEntry } from "./types";

/** Cache TTL in milliseconds. 7 days, per the master plan. */
export const CACHE_TTL_MS = 7 * 24 * 60 * 60 * 1000;

/** Default on-disk cache path. */
export function defaultCachePath(): string {
  return path.join(os.homedir(), ".subscription-service", "license-cache.json");
}

/** Read the cache. Returns `null` when the file does not exist or cannot be parsed. */
export async function readCache(cachePath: string = defaultCachePath()): Promise<LicenseCacheEntry | null> {
  let raw: string;
  try {
    raw = await fs.readFile(cachePath, "utf-8");
  } catch (err) {
    if (isNodeNotFound(err)) return null;
    // Other read errors (permissions, etc.) are treated like a missing cache;
    // a corrupt cache shouldn't crash UI boot.
    return null;
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    return null;
  }

  if (!isCacheEntry(parsed)) return null;
  return parsed;
}

/** Write the cache atomically (write tmp, rename) so a partial write can never corrupt it. */
export async function writeCache(
  entry: LicenseCacheEntry,
  cachePath: string = defaultCachePath(),
): Promise<void> {
  const dir = path.dirname(cachePath);
  await fs.mkdir(dir, { recursive: true });
  const tmp = `${cachePath}.tmp-${process.pid}-${Date.now()}`;
  await fs.writeFile(tmp, JSON.stringify(entry, null, 2), { encoding: "utf-8", mode: 0o600 });
  await fs.rename(tmp, cachePath);
}

/**
 * Returns the cached entry IF it exists AND is fresh (fetchedAt within
 * CACHE_TTL_MS of `now`). Returns `null` otherwise. Callers that still want
 * the stale entry (e.g. for diagnostics) should call `readCache` directly.
 */
export async function readFreshCache(
  cachePath: string = defaultCachePath(),
  now: Date = new Date(),
): Promise<LicenseCacheEntry | null> {
  const entry = await readCache(cachePath);
  if (!entry) return null;
  if (!isFresh(entry, now)) return null;
  return entry;
}

/** True iff `fetchedAt` is within CACHE_TTL_MS of `now`. */
export function isFresh(entry: LicenseCacheEntry, now: Date = new Date()): boolean {
  const fetchedAt = Date.parse(entry.fetchedAt);
  if (Number.isNaN(fetchedAt)) return false;
  return now.getTime() - fetchedAt < CACHE_TTL_MS;
}

/** When the entry's cache window expires. */
export function cacheValidUntil(entry: LicenseCacheEntry): Date {
  return new Date(Date.parse(entry.fetchedAt) + CACHE_TTL_MS);
}

/** Build a cache entry from a fresh server response and the current time. */
export function buildCacheEntry(
  response: EntitlementResponse,
  licenseKeyFingerprint: string,
  fetchedAt: Date = new Date(),
): LicenseCacheEntry {
  return {
    fetchedAt: fetchedAt.toISOString(),
    licenseKeyFingerprint,
    response,
  };
}

// ---------------------------------------------------------------------------

function isNodeNotFound(err: unknown): boolean {
  return typeof err === "object" && err !== null && (err as { code?: string }).code === "ENOENT";
}

function isCacheEntry(value: unknown): value is LicenseCacheEntry {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  if (typeof v.fetchedAt !== "string") return false;
  if (typeof v.licenseKeyFingerprint !== "string") return false;
  if (typeof v.response !== "object" || v.response === null) return false;
  const r = v.response as Record<string, unknown>;
  if (!Array.isArray(r.entitlements)) return false;
  if (!r.entitlements.every((e) => typeof e === "string")) return false;
  if (typeof r.expiresAt !== "string") return false;
  if (typeof r.tierName !== "string") return false;
  if (typeof r.signature !== "string") return false;
  return true;
}

File: interface-engine/src/main/kotlin/com/bzonfhir/subscriptionservice/interfaceengine/entitlements/LicenseCache.kt
package com.bzonfhir.subscriptionservice.interfaceengine.entitlements

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import org.slf4j.LoggerFactory
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.time.Duration
import java.time.Instant

/**
 * On-disk license cache with a 7-day TTL (ticket #460, Epic #428).
 *
 * JVM mirror of `ui/src/lib/license/licenseCache.ts`. Why disk? A
 * license-server outage MUST NOT degrade the engine all the way to FOSS the
 * moment it starts up if it has a fresh cached set on disk. The engine reads
 * the cache at startup AND on every refresh tick (every 12h by default); a
 * transient outage degrades to "stale-active" rather than "FOSS". After 7
 * days without a successful fetch we give up and fall back to FOSS so a
 * permanently-misconfigured server doesn't pretend forever.
 *
 * Path: `~/.subscription-service/license-cache.json` by default; the
 * EntitlementHolder reads `subscription-service.license.cache-path` from
 * config and passes it down.
 *
 * Atomicity: writes go to a `<path>.tmp-<pid>-<ts>` sibling first, then
 * `Files.move(REPLACE_EXISTING, ATOMIC_MOVE)`. A partial write can never
 * leave the cache corrupt — a reader either sees the old file or the new
 * file, never an in-flight truncation.
 *
 * Stateless: this object is a pair of pure functions over a path. All
 * Spring-managed lifecycle lives in EntitlementHolder.
 */
object LicenseCache {

    /** Cache freshness window. Matches the UI's `CACHE_TTL_MS = 7 * 24 * 60 * 60 * 1000`. */
    val TTL: Duration = Duration.ofDays(7)

    private val log = LoggerFactory.getLogger(LicenseCache::class.java)

    private val mapper: ObjectMapper = ObjectMapper()
        .registerModule(JavaTimeModule())
        // Instant -> "2026-06-28T00:00:00Z" rather than 1_751_068_800 — match the UI's JSON shape.
        .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
        // An older cache from a prior epic might have fewer fields; ignore them rather than crash boot.
        .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
        .setSerializationInclusion(JsonInclude.Include.NON_NULL)

    /** Read the cache. Returns `null` on missing file, IO error, or malformed JSON. */
    fun read(cachePath: String): LicenseCacheEntry? {
        val path = Path.of(cachePath)
        if (!Files.exists(path)) return null
        return try {
            val bytes = Files.readAllBytes(path)
            mapper.readValue(bytes, LicenseCacheEntry::class.java)
        } catch (e: Exception) {
            // A corrupt cache shouldn't crash boot. Log + treat as missing.
            log.warn("failed to read license cache at {} — treating as absent: {}", cachePath, e.message)
            null
        }
    }

    /** Write the cache atomically. Creates the parent directory if necessary. */
    fun write(entry: LicenseCacheEntry, cachePath: String) {
        val path = Path.of(cachePath)
        val parent = path.parent
        if (parent != null) Files.createDirectories(parent)
        val tmp = path.resolveSibling("${path.fileName}.tmp-${ProcessHandle.current().pid()}-${System.currentTimeMillis()}")
        Files.write(tmp, mapper.writerWithDefaultPrettyPrinter().writeValueAsBytes(entry))
        // ATOMIC_MOVE so a reader can't observe a half-written file.
        Files.move(tmp, path, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE)
    }

    /** True iff `fetchedAt` is within [TTL] of [now]. */
    fun isFresh(entry: LicenseCacheEntry, now: Instant = Instant.now()): Boolean {
        return Duration.between(entry.fetchedAt, now) < TTL
    }
}

/**
 * On-disk cache entry. Mirrors the UI's `LicenseCacheEntry` shape so that
 * future tooling (e.g. a "view cache" admin endpoint) can read the same
 * JSON on both sides.
 */
data class LicenseCacheEntry @JsonCreator constructor(
    /** When the response was fetched. ISO-8601 in the on-disk JSON. */
    @param:JsonProperty("fetchedAt") val fetchedAt: Instant,
    /** First 8 chars of sha256(licenseKey). Safe to log. */
    @param:JsonProperty("licenseKeyFingerprint") val licenseKeyFingerprint: String,
    /** The verified entitlement strings. */
    @param:JsonProperty("entitlements") val entitlements: List<String>,
    /** Tier name from the server. */
    @param:JsonProperty("tierName") val tierName: String,
    /** License-expiry timestamp (ISO-8601 string as the server sent it). */
    @param:JsonProperty("expiresAt") val expiresAt: String,
    /** Compact JWS of the entitlement body. */
    @param:JsonProperty("signature") val signature: String,
    /** Kid of the JWK that verified this entry. */
    @param:JsonProperty("signatureKid") val signatureKid: String? = null,
)
